### PR TITLE
feature/datadog-stats

### PR DESCRIPTION
**Pull request by johantjellden**

### CODE REVIEW

File: member-service/src/main/java/com/hedvig/memberservice/jobs/BankidMetrics.kt [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package com.hedvig.memberservice.jobs
+
+import io.micrometer.core.instrument.MeterRegistry
+import org.springframework.stereotype.Component
+
+@Component
+class BankidMetrics(
```
Comment (FredrikAreschoug) [inline member-service/src/main/java/com/hedvig/memberservice/jobs/BankidMetrics.kt:7]:
```suggestion
class SwedishBankidMetrics(
```

File: member-service/src/main/java/com/hedvig/memberservice/jobs/BankidMetrics.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    registry: MeterRegistry) {
+
+    private val authStartedCounter = registry.counter("bankid.auth.started.count")
+    private val authFailedCounter =  registry.counter("bankid.auth.failed.count")
+    private val authCompletedCounter =  registry.counter("bankid.auth.completed.count")
+
+    private val signStartedCounter = registry.counter("bankid.sign.started.count")
+    private val signFailedCounter = registry.counter("bankid.sign.failed.count")
+    private val signCompletedCounter = registry.counter("bankid.sign.completed.count")
```
Comment (FredrikAreschoug) [inline member-service/src/main/java/com/hedvig/memberservice/jobs/BankidMetrics.kt:16]: Should we add `swedish.` to this?

**Review by FredrikAreschoug — COMMENTED** [review]
Looks good, but I think we should make clear that this is only an implementation for sweden

File: member-service/src/main/java/com/hedvig/memberservice/jobs/BankidMetrics.kt (hunk 1 above)
Comment (vonElfvin) [reply member-service/src/main/java/com/hedvig/memberservice/jobs/BankidMetrics.kt:7]: Also, small I on Id/ID looks weird
```suggestion
class SwedishBankIdMetrics(
```

File: member-service/src/main/java/com/hedvig/memberservice/jobs/BankidMetrics.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        signStartedCounter.increment()
+    }
+
+    fun completeBankIdSign() {
+        signCompletedCounter
+    }
+
+    fun failBankIdSign() {
+        signFailedCounter.increment()
+    }
```
Comment (vonElfvin) [inline member-service/src/main/java/com/hedvig/memberservice/jobs/BankidMetrics.kt:39]: Are the non V2 even used anymore? Or is this just a way to see that?

File: member-service/src/main/java/com/hedvig/memberservice/services/SwedishBankIdSigningService.kt [hunk 4]
```diff
@@ -33,7 +34,8 @@ class SwedishBankIdSigningService(
     @Value("\${hedvig.bankid.signmessage.switcher}")
     private val switcherMessage: String,
     @Value("\${hedvig.bankid.signmessage.nonSwitcher}")
-    private val nonSwitcherMessage: String
+    private val nonSwitcherMessage: String,
+    private val bankidMetrics: BankidMetrics
```
Comment (vonElfvin) [inline member-service/src/main/java/com/hedvig/memberservice/services/SwedishBankIdSigningService.kt:38]:
```suggestion
    private val bankIdMetrics: BankidMetrics
```

File: member-service/src/main/java/com/hedvig/memberservice/services/v2/BankIdServiceV2.kt [hunk 5]
```diff
@@ -31,7 +32,8 @@ class BankIdServiceV2(
     private val redisEventPublisher: RedisEventPublisher,
     private val scheduler: Scheduler,
     private val collectRepository: CollectRepository,
-    private val apiGatewayService: ApiGatewayService
+    private val apiGatewayService: ApiGatewayService,
```
Comment (vonElfvin) [inline member-service/src/main/java/com/hedvig/memberservice/services/v2/BankIdServiceV2.kt:35]: Maybe rename the file ☝️ to `SwedishBankIdServiceV2` or preferably remove the non V2 and keep this one as `SwedishBankIdService`

**Review by vonElfvin — COMMENTED** [review]
Cool! I agree with Fredrik on the "Swedish" naming, on that note, let's add Norway/Denmark while we're at it no?

File: member-service/src/main/java/com/hedvig/memberservice/services/v2/BankIdServiceV2.kt (hunk 5 above)
Comment (johantjellden) [reply member-service/src/main/java/com/hedvig/memberservice/services/v2/BankIdServiceV2.kt:35]: Removing the old service feels a bit like scope creep. The old service is still used so lets do that another time.

**Review by FredrikAreschoug — APPROVED** [review]
:100:

File: member-service/src/main/java/com/hedvig/memberservice/services/v2/BankIdServiceV2.kt (hunk 5 above)
Comment (vonElfvin) [reply member-service/src/main/java/com/hedvig/memberservice/services/v2/BankIdServiceV2.kt:35]: Yea I figured if it wasn't used. But if it's still used then let's keep it of course :)

**Review by vonElfvin — APPROVED** [review]

File: zign-sec/src/main/java/com/hedvig/external/Metrics.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    private val sessionFailedCounter = generateMethodCounters("zignSec.session.fail")
+    private val sessionSuccessCounter = generateMethodCounters("zignSec.session.success")
+    private val sessionStartedCounter = generateMethodCounters("zignSec.session.started")
+
+    fun authRequestFailed(request: ZignSecBankIdAuthenticationRequest, type: ZignSecAuthenticationType) {
+        authRequestFailedCounters[request.authMethod]?.get(type)?.increment()
+    }
+
+    fun authRequestSuccess(authMethod: ZignSecAuthenticationMethod, type: ZignSecAuthenticationType) {
+        sessionStartedCounter[authMethod]?.get(type)?.increment()
```
Comment (FredrikAreschoug) [inline zign-sec/src/main/java/com/hedvig/external/Metrics.kt:24]: I can only see that we use the start counter here shouldn't we do it in the beginning of startNewSession?

File: zign-sec/src/main/java/com/hedvig/external/Metrics.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        authRequestSuccessCounters[authMethod]?.get(type)?.increment()
+    }
+
+    fun signSessionFailed(authenticationMethod: ZignSecAuthenticationMethod, requestType: ZignSecAuthenticationType) {
```
Comment (FredrikAreschoug) [inline zign-sec/src/main/java/com/hedvig/external/Metrics.kt:28]: Should we rename this one to authSessionFailed? hence it is used in both Authentication and Signing :slightly_smiling_face:

File: zign-sec/src/main/java/com/hedvig/external/Metrics.kt [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        sessionFailedCounter[authenticationMethod]?.get(requestType)?.increment()
+    }
+
+    fun signSessionSuccess(authenticationMethod: ZignSecAuthenticationMethod, requestType: ZignSecAuthenticationType) {
```
Comment (FredrikAreschoug) [inline zign-sec/src/main/java/com/hedvig/external/Metrics.kt:32]: Same here

**Review by FredrikAreschoug — COMMENTED** [review]
Looks good just some naming and a question on the start counter

**Review by FredrikAreschoug — APPROVED** [review]
:taco:

**Pull request merged**